## Introduction 

Repeat customers are the most valuable customers you have.

Repeat customers:

- Are loyal and satisfied
- Make multiple purchases
- Talk about your brand
- Refer your brand to their friends

However, maintaining your relationship with loyal customers can be tricky.These relationships need to be maintained with a consistent and rewarding customer experience. When your customers are happy, your business will prosper.

What is Churn ?
Customers that have churned are customers that have cut ties with a business or brand.

What is Churn Rate used for ?
Churn rate is a measurement of the health of your business.
Churn measures:

- The success of increasing customer retention for a company
- The changes that affect customer retention
- Customer Lifetime Value
- The type of customers that have the highest retention rate with your company
- Much more!

Today, we are going to analyse a customer dataset of a bank to provide them with valuable insights for their ongoing customer retention issue. We would apply the ANN(Artificial Neural Networks) to classify the dataset and provide the bank with the best fit model. The problem statement is mentioned in detail below.

## Problem Statement

The bank has been seen unusual churn rates for their customers (churn is when people leave the company). They want to understand the problem for this unusual high churn rates. Here we have been provided with a sample of their, last six months customers whose charecteristics like 

- credit score 
- geography
- gender
- age
- tenure (how many years a customer is with the bank)
- balance
- number of products a customer had with the bank
- if he/she had a credit card
- if he/she is an active member 
- estimated salary (the bank has estimated the salary based of the data they had)
- exited (if a customer has left the bank or not within the last six months)  

Our task is to create a classification model for predicting customers at risk of churning. From this classification model the bank could gain useful insights and can take relevant actions to prevent further unusual churn rate and improve the overall customer experience.

## Data Preprossesing  

In [89]:
# Importing the required libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [90]:
#Reading the dataset
dataset = pd.read_csv('/Users/Lenovo/Desktop/Churn_Modelling.csv')
dataset

RowNumber  CustomerId         Surname  CreditScore Geography  Gender  \
0             1    15634602        Hargrave          619    France  Female   
1             2    15647311            Hill          608     Spain  Female   
2             3    15619304            Onio          502    France  Female   
3             4    15701354            Boni          699    France  Female   
4             5    15737888        Mitchell          850     Spain  Female   
5             6    15574012             Chu          645     Spain    Male   
6             7    15592531        Bartlett          822    France    Male   
7             8    15656148          Obinna          376   Germany  Female   
8             9    15792365              He          501    France    Male   
9            10    15592389              H?          684    France    Male   
10           11    15767821          Bearce          528    France    Male   
11           12    15737173         Andrews          497     Spain    Male   
12           13    15632264             Kay          476    France  Female   
13           14    15691483            Chin          549    France  Female   
14           15    15600882           Scott          635     Spain  Female   
15           16    15643966         Goforth          616   Germany    Male   
16           17    15737452           Romeo          653   Germany    Male   
17           18    15788218       Henderson          549     Spain  Female   
18           19    15661507         Muldrow          587     Spain    Male   
19           20    15568982             Hao          726    France  Female   
20           21    15577657        McDonald          732    France    Male   
21           22    15597945        Dellucci          636     Spain  Female   
22           23    15699309       Gerasimov          510     Spain  Female   
23           24    15725737          Mosman          669    France    Male   
24           25    15625047             Yen          846    France  Female   
25           26    15738191         Maclean          577    France    Male   
26           27    15736816           Young          756   Germany    Male   
27           28    15700772         Nebechi          571    France    Male   
28           29    15728693      McWilliams          574   Germany  Female   
29           30    15656300        Lucciano          411    France    Male   
...         ...         ...             ...          ...       ...     ...   
9970       9971    15587133        Thompson          518    France    Male   
9971       9972    15721377            Chou          833    France  Female   
9972       9973    15747927           Ch'in          758    France    Male   
9973       9974    15806455          Miller          611    France    Male   
9974       9975    15695474          Barker          583    France    Male   
9975       9976    15666295           Smith          610   Germany    Male   
9976       9977    15656062         Azikiwe          637    France  Female   
9977       9978    15579969         Mancini          683    France  Female   
9978       9979    15703563           P'eng          774    France    Male   
9979       9980    15692664          Diribe          677    France  Female   
9980       9981    15719276            T'ao          741     Spain    Male   
9981       9982    15672754        Burbidge          498   Germany    Male   
9982       9983    15768163         Griffin          655   Germany  Female   
9983       9984    15656710           Cocci          613    France    Male   
9984       9985    15696175  Echezonachukwu          602   Germany    Male   
9985       9986    15586914          Nepean          659    France    Male   
9986       9987    15581736        Bartlett          673   Germany    Male   
9987       9988    15588839         Mancini          606     Spain    Male   
9988       9989    15589329         Pirozzi          775    France    Male   
9989       9990    15605622 

In [91]:
#Summary of the dataset 
dataset.describe()

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

We can see that their is no missing value in the dataset as the size of the dataset is 10000x14 and all the attributes have a count of 10000.  

We do not consider the row number, the customer id and the surname as independent variables as they would have no impact on our dependent variable. Hence we trim the dataset and divide the dataset as follows: 

In [92]:
X = dataset.iloc[:, 3:13].values #Feature Variable 
y = dataset.iloc[:, 13].values #Target Variable

In [93]:
print(X)
print(y)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]
[1 0 1 ... 1 1 0]


Our model needs in input numerical data, so, we need to encode categorical data into numerical data. In This case we have Geography (France, Spain and Germany) and Gender (Male and Female). For Geography we will have 0,1,2 instead of France, Spain and Gemany and 0,1 instead of Gender.

In [94]:
#Converting to Variables 
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])

We have a binary value for the column 'Gender' hence we don't have issues with the encoding. 

But, for the column 'Geography', we have 3 possible values and considering that our categorical values have no comparison (for example, Germany is not higher than France), we need to create dummy variables, so we will create 3 columns, one for each country where we will have 1 if the customer is from that country and 0 if not. And to avoid the dummy variable trap, we will also delete 1 of the 3 columns:

In [95]:
#Creating a dummy variable 
onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:]

We now split the dataset into training and testing sets.

In [96]:
#Splitting the dataset into Training and testing sets 
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

We need to apply feature scaling to the dataset to standardize the dataset to convert them into same sclae for faster computation of results.

In [97]:
#Feature scaling 
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Artificial Neural Network Clustering 

In [98]:
#Importing the required libraries
import keras
from keras.models import Sequential
from keras.layers import Dense

### Model 1 

We now create the ANN with one hidden layer and having 6 nodes (Half of feature and target variables) with activation function of the hidden and the input layer as 'relu' (rectified linear units) and the output layer 'sigmoid' as we need binary result for our classification model. Then we compile the model with SGD 'adam' as its the best which can be used in thos case. 

In [99]:
#Creating a Neural network 
classifier = Sequential()
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu', input_dim = 11))#The Input layer 
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu'))#The hidden layer 1
classifier.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid'))#The Output layer 
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])#Compiling the result with SGD as adam

C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=11, units=6, kernel_initializer="uniform")`
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", units=6, kernel_initializer="uniform")`
  after removing the cwd from sys.path.
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="sigmoid", units=1, kernel_initializer="uniform")`
  """


In [100]:
#Fitting the Neural network model on the testing data 
classifier.fit(X_train, y_train, batch_size = 10, nb_epoch = 100)

C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  


Epoch 1/100
8000/8000 [==============================] - 4s 499us/step - loss: 0.4931 - acc: 0.7954
Epoch 2/100
8000/8000 [==============================] - 2s 265us/step - loss: 0.4302 - acc: 0.7960
Epoch 3/100
8000/8000 [==============================] - 2s 255us/step - loss: 0.4255 - acc: 0.7960
Epoch 4/100
8000/8000 [==============================] - 2s 221us/step - loss: 0.4210 - acc: 0.7999
Epoch 5/100
8000/8000 [==============================] - 2s 233us/step - loss: 0.4182 - acc: 0.8226
Epoch 6/100
8000/8000 [==============================] - 2s 220us/step - loss: 0.4161 - acc: 0.8262
Epoch 7/100
8000/8000 [==============================] - 2s 228us/step - loss: 0.4144 - acc: 0.8294
Epoch 8/100
8000/8000 [==============================] - 1s 184us/step - loss: 0.4132 - acc: 0.8292
Epoch 9/100
8000/8000 [==============================] - 1s 178us/step - loss: 0.4125 - acc: 0.8312
Epoch 10/100
8000/8000 [==============================] - 2s 200us/step - loss: 0.4113 - acc: 0.8317

8000/8000 [==============================] - 2s 199us/step - loss: 0.4003 - acc: 0.8362
Epoch 83/100
8000/8000 [==============================] - 2s 218us/step - loss: 0.4003 - acc: 0.8349
Epoch 84/100
8000/8000 [==============================] - 2s 201us/step - loss: 0.4002 - acc: 0.8355
Epoch 85/100
8000/8000 [==============================] - 1s 178us/step - loss: 0.3999 - acc: 0.8346
Epoch 86/100
8000/8000 [==============================] - 1s 179us/step - loss: 0.4000 - acc: 0.8341
Epoch 87/100
8000/8000 [==============================] - 2s 196us/step - loss: 0.3997 - acc: 0.8364 0s - loss:
Epoch 88/100
8000/8000 [==============================] - 2s 239us/step - loss: 0.4002 - acc: 0.8355
Epoch 89/100
8000/8000 [==============================] - 2s 213us/step - loss: 0.3996 - acc: 0.8359
Epoch 90/100
8000/8000 [==============================] - 2s 208us/step - loss: 0.3996 - acc: 0.8342
Epoch 91/100
8000/8000 [==============================] - 2s 209us/step - loss: 0.3996 - acc:

In [101]:
#Predicting the testing set results
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)

In [102]:
#Creating the confusion matrix 
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[1539   56]
 [ 256  149]]


In [103]:
#Calculating the accuracy of the model
accuracy_ANN = (cm[0,0]+cm[1,1])/ (cm[0,0]+cm[0,1]+cm[1,0]+cm[1,1])
accuracy_ANN

0.844

What is Specificity ?

Specificity (also called the true negative rate) measures the proportion of actual negatives that are correctly identified.
Specificity relates to the test's ability to correctly reject customers who did not churn and actually did'nt churn.

In [104]:
#Calculating the specificity of the model
specificity_ANN =(cm[1,1]/(cm[1,1]+cm[0,1]))
specificity_ANN

0.7268292682926829

What is Sensitivity ?

Sensitivity (also called the true positive rate) measures the proportion of actual positives that are correctly identified. Sensitivity refers to the test's ability to correctly detect people who are going to churn and actually do churn.

In [105]:
#Calculating the sensitivity of the model
sensitivity_ANN =(cm[0,0]/(cm[0,0]+cm[1,0]))
sensitivity_ANN

0.8573816155988858

In [125]:
#loss function 
classifier_1.evaluate(X_test,y_pred1, verbose=1)

2000/2000 [==============================] - 0s 38us/step


[0.17338521599769594, 1.0]

### Model 2

We now create the ANN with one hidden layer and having 6 nodes (Half of feature and target variables) with activation function of the hidden and the input layer as 'tanh' (hyperbolic tangent) and the output layer 'sigmoid' as we need binary result for our classification model. Then we compile the model with SGD 'adam' as its the best which can be used in this case.

In [106]:
#Creating a Neural Network model 
classifier_1 = Sequential()
classifier_1.add(Dense(output_dim = 6, init = 'uniform', activation = 'tanh', input_dim = 11))
classifier_1.add(Dense(output_dim = 6, init = 'uniform', activation = 'tanh'))
classifier_1.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid'))
classifier_1.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="tanh", input_dim=11, units=6, kernel_initializer="uniform")`
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="tanh", units=6, kernel_initializer="uniform")`
  after removing the cwd from sys.path.
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="sigmoid", units=1, kernel_initializer="uniform")`
  """


In [107]:
#Fitting the model on testing data 
classifier_1.fit(X_train, y_train, batch_size = 10, nb_epoch = 100)

C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  


Epoch 1/100
8000/8000 [==============================] - 3s 344us/step - loss: 0.4769 - acc: 0.8085
Epoch 2/100
8000/8000 [==============================] - 2s 212us/step - loss: 0.4335 - acc: 0.8121
Epoch 3/100
8000/8000 [==============================] - 2s 264us/step - loss: 0.4322 - acc: 0.8102
Epoch 4/100
8000/8000 [==============================] - 2s 258us/step - loss: 0.4293 - acc: 0.8116
Epoch 5/100
8000/8000 [==============================] - 2s 241us/step - loss: 0.4202 - acc: 0.8144
Epoch 6/100
8000/8000 [==============================] - 2s 213us/step - loss: 0.4023 - acc: 0.8306
Epoch 7/100
8000/8000 [==============================] - 2s 217us/step - loss: 0.3820 - acc: 0.8391
Epoch 8/100
8000/8000 [==============================] - 2s 221us/step - loss: 0.3660 - acc: 0.8510
Epoch 9/100
8000/8000 [==============================] - 2s 218us/step - loss: 0.3565 - acc: 0.8545
Epoch 10/100
8000/8000 [==============================] - 2s 242us/step - loss: 0.3520 - acc: 0.8579

8000/8000 [==============================] - 2s 247us/step - loss: 0.3327 - acc: 0.8611
Epoch 82/100
8000/8000 [==============================] - 2s 209us/step - loss: 0.3325 - acc: 0.8639
Epoch 83/100
8000/8000 [==============================] - 2s 222us/step - loss: 0.3328 - acc: 0.8631
Epoch 84/100
8000/8000 [==============================] - 2s 224us/step - loss: 0.3318 - acc: 0.8665
Epoch 85/100
8000/8000 [==============================] - 2s 243us/step - loss: 0.3328 - acc: 0.8631
Epoch 86/100
8000/8000 [==============================] - 2s 221us/step - loss: 0.3323 - acc: 0.8635
Epoch 87/100
8000/8000 [==============================] - 2s 215us/step - loss: 0.3325 - acc: 0.8637
Epoch 88/100
8000/8000 [==============================] - 2s 269us/step - loss: 0.3321 - acc: 0.8654
Epoch 89/100
8000/8000 [==============================] - 2s 251us/step - loss: 0.3328 - acc: 0.8629
Epoch 90/100
8000/8000 [==============================] - 2s 240us/step - loss: 0.3326 - acc: 0.8630
Epo

In [108]:
#Predicting the target values using the model on testing data 
y_pred1 = classifier_1.predict(X_test)
y_pred1 = (y_pred1 > 0.5)

In [109]:
#Creating the confusion matrix 
from sklearn.metrics import confusion_matrix
cm1 = confusion_matrix(y_test, y_pred1)
print(cm1)

[[1513   82]
 [ 194  211]]


In [110]:
#Calculating the accuracy of the model
accuracy_ANN1 = (cm1[0,0]+cm1[1,1])/ (cm1[0,0]+cm1[0,1]+cm1[1,0]+cm1[1,1])
accuracy_ANN1

0.862

In [111]:
#Calculating the specificity of the model
specificity_ANN1 =(cm1[1,1]/(cm1[1,1]+cm1[0,1]))
specificity_ANN1

0.7201365187713311

In [112]:
#Calculating the sensitivity of the model
sensitivity_ANN1 =(cm1[0,0]/(cm1[0,0]+cm1[1,0]))
sensitivity_ANN1

0.8863503222026948

In [113]:
#Calculating the loss function of the model 
classifier_1.evaluate(X_test,y_pred1, verbose=1)

2000/2000 [==============================] - 0s 181us/step


[0.17338521599769594, 1.0]

### Model 3 

We now create the ANN with two hidden layer and having 6 nodes (Half of feature and target variables) with activation function of the hidden and the input layers as 'tanh' (hyperbolic tangent) and the output layer 'sigmoid' as we need binary result for our classification model. Then we compile the model with SGD 'adam' as its the best which can be used in this case.

In [115]:
#Creating a Neural Network model 
classifier_2 = Sequential()
classifier_2.add(Dense(output_dim = 6, init = 'uniform', activation = 'tanh', input_dim = 11))
classifier_2.add(Dense(output_dim = 6, init = 'uniform', activation = 'tanh'))
classifier_2.add(Dense(output_dim = 6, init = 'uniform', activation = 'tanh'))
classifier_2.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid'))
classifier_2.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
classifier_2.fit(X_train, y_train, batch_size = 10, nb_epoch = 100)

C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="tanh", input_dim=11, units=6, kernel_initializer="uniform")`
  
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="tanh", units=6, kernel_initializer="uniform")`
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="tanh", units=6, kernel_initializer="uniform")`
  after removing the cwd from sys.path.
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="sigmoid", units=1, kernel_initializer="uniform")`
  """
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: UserWarning

Epoch 1/100
8000/8000 [==============================] - 8s 986us/step - loss: 0.4787 - acc: 0.8046
Epoch 2/100
8000/8000 [==============================] - 2s 272us/step - loss: 0.4353 - acc: 0.8164
Epoch 3/100
8000/8000 [==============================] - 2s 228us/step - loss: 0.4341 - acc: 0.8129
Epoch 4/100
8000/8000 [==============================] - 2s 227us/step - loss: 0.4310 - acc: 0.8129
Epoch 5/100
8000/8000 [==============================] - 2s 223us/step - loss: 0.4257 - acc: 0.8171
Epoch 6/100
8000/8000 [==============================] - 2s 232us/step - loss: 0.4179 - acc: 0.8239
Epoch 7/100
8000/8000 [==============================] - 2s 237us/step - loss: 0.4103 - acc: 0.8301
Epoch 8/100
8000/8000 [==============================] - 2s 231us/step - loss: 0.3993 - acc: 0.8360
Epoch 9/100
8000/8000 [==============================] - 2s 286us/step - loss: 0.3805 - acc: 0.8436
Epoch 10/100
8000/8000 [==============================] - 2s 285us/step - loss: 0.3644 - acc: 0.8511

8000/8000 [==============================] - 2s 241us/step - loss: 0.3348 - acc: 0.8627
Epoch 82/100
8000/8000 [==============================] - 2s 275us/step - loss: 0.3348 - acc: 0.8627
Epoch 83/100
8000/8000 [==============================] - 2s 277us/step - loss: 0.3355 - acc: 0.8626
Epoch 84/100
8000/8000 [==============================] - 2s 235us/step - loss: 0.3353 - acc: 0.8631
Epoch 85/100
8000/8000 [==============================] - 2s 226us/step - loss: 0.3339 - acc: 0.8641
Epoch 86/100
8000/8000 [==============================] - 2s 252us/step - loss: 0.3343 - acc: 0.8652
Epoch 87/100
8000/8000 [==============================] - 2s 231us/step - loss: 0.3348 - acc: 0.8634
Epoch 88/100
8000/8000 [==============================] - 2s 231us/step - loss: 0.3331 - acc: 0.8636
Epoch 89/100
8000/8000 [==============================] - 2s 254us/step - loss: 0.3347 - acc: 0.8627
Epoch 90/100
8000/8000 [==============================] - 2s 281us/step - loss: 0.3341 - acc: 0.8651
Epo

In [116]:
#Predicting the model with test set 
y_pred2 = classifier_2.predict(X_test)
y_pred2 = (y_pred2 > 0.5)
y_pred2

array([[False],
       [False],
       [False],
       ...,
       [False],
       [False],
       [False]])

In [117]:
#Creating the confusion matrix 
cm2 = confusion_matrix(y_test, y_pred2)
print(cm2)

#Accuracy of the model 
accuracy_ANN2 = (cm2[0,0]+cm2[1,1])/ (cm2[0,0]+cm2[0,1]+cm2[1,0]+cm2[1,1])
accuracy_ANN2

[[1521   74]
 [ 203  202]]


0.8615

In [118]:
#Calculating the specificity of the model
specificity_ANN2 =(cm2[1,1]/(cm2[1,1]+cm2[0,1]))
specificity_ANN2

0.7318840579710145

In [119]:
#Calculating the sensitivity of the model
sensitivity_ANN2 =(cm2[0,0]/(cm2[0,0]+cm2[1,0]))
sensitivity_ANN2

0.8822505800464037

### Model 4

In [120]:
#Creating the Neural Network model 
classifier_2 = Sequential()
classifier_2.add(Dense(output_dim = 6, init = 'uniform', activation = 'tanh', input_dim = 11))
classifier_2.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu'))
classifier_2.add(Dense(output_dim = 6, init = 'uniform', activation = 'tanh'))
classifier_2.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid'))
classifier_2.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

#Fitting the model on testing data 
classifier_2.fit(X_train, y_train, batch_size = 10, nb_epoch = 100)

C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="tanh", input_dim=11, units=6, kernel_initializer="uniform")`
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", units=6, kernel_initializer="uniform")`
  after removing the cwd from sys.path.
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="tanh", units=6, kernel_initializer="uniform")`
  """
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="sigmoid", units=1, kernel_initializer="uniform")`
  
C:\Users\Lenovo\Anaconda3\lib\site-packages\ipykernel_launcher.py:10: UserWarnin

Epoch 1/100
8000/8000 [==============================] - 3s 393us/step - loss: 0.4793 - acc: 0.7941
Epoch 2/100
8000/8000 [==============================] - 2s 249us/step - loss: 0.4314 - acc: 0.7951
Epoch 3/100
8000/8000 [==============================] - 2s 261us/step - loss: 0.4301 - acc: 0.7974
Epoch 4/100
8000/8000 [==============================] - 2s 233us/step - loss: 0.4278 - acc: 0.8022
Epoch 5/100
8000/8000 [==============================] - 2s 270us/step - loss: 0.4226 - acc: 0.8131
Epoch 6/100
8000/8000 [==============================] - 2s 282us/step - loss: 0.4153 - acc: 0.8211
Epoch 7/100
8000/8000 [==============================] - 2s 273us/step - loss: 0.4006 - acc: 0.8326
Epoch 8/100
8000/8000 [==============================] - 2s 234us/step - loss: 0.3702 - acc: 0.8511
Epoch 9/100
8000/8000 [==============================] - 2s 243us/step - loss: 0.3553 - acc: 0.8579
Epoch 10/100
8000/8000 [==============================] - 2s 243us/step - loss: 0.3501 - acc: 0.8595

8000/8000 [==============================] - 2s 274us/step - loss: 0.3333 - acc: 0.8634
Epoch 83/100
8000/8000 [==============================] - 2s 253us/step - loss: 0.3327 - acc: 0.8639
Epoch 84/100
8000/8000 [==============================] - 2s 274us/step - loss: 0.3334 - acc: 0.8640
Epoch 85/100
8000/8000 [==============================] - 2s 281us/step - loss: 0.3327 - acc: 0.8632
Epoch 86/100
8000/8000 [==============================] - 2s 236us/step - loss: 0.3328 - acc: 0.8625
Epoch 87/100
8000/8000 [==============================] - 2s 226us/step - loss: 0.3323 - acc: 0.8634
Epoch 88/100
8000/8000 [==============================] - 2s 221us/step - loss: 0.3324 - acc: 0.8647
Epoch 89/100
8000/8000 [==============================] - 2s 243us/step - loss: 0.3317 - acc: 0.8646
Epoch 90/100
8000/8000 [==============================] - 2s 240us/step - loss: 0.3325 - acc: 0.8630
Epoch 91/100
8000/8000 [==============================] - 2s 263us/step - loss: 0.3322 - acc: 0.8656
Epo

In [121]:
#Predicting the target value using the model on testing data 
y_pred3 = classifier_2.predict(X_test)
y_pred3 = (y_pred3 > 0.5)

In [122]:
#Creating the confusion matrix 
cm3 = confusion_matrix(y_test, y_pred3)
print(cm3)
#Calculating the accuracy of the model 
accuracy_ANN3 = (cm3[0,0]+cm3[1,1])/ (cm3[0,0]+cm3[0,1]+cm3[1,0]+cm3[1,1])
accuracy_ANN3

[[1515   80]
 [ 201  204]]


0.8595

In [123]:
#Calculating the Specificity of the model  
specificity_ANN3 =(cm3[1,1]/(cm3[1,1]+cm3[0,1]))
specificity_ANN3

0.7183098591549296

In [124]:
##Calculating the Sensitivity of the model
sensitivity_ANN3 =(cm3[0,0]/(cm3[0,0]+cm3[1,0]))
sensitivity_ANN3

0.8828671328671329

## Conclusion 

The model 2 has the accurate representation of the churn rate of bank customers as we achieved highest accuracy among all the ANN models with varying activation functions and number of hidden layers. It has better accuracy of 86.2% with specificity 72.01% and 88.6% sensitivity and low loss function of 0.173.

The model 2 can provide most accurate classification to the bank for their customers at risk of churning. From that data the bank can analyse the customer's requirement and problems to cater the customers in a better manner and to avoid churn.  